filters/auth/oauth.go

```diff
@@ -0,0 +1,429 @@
+	AuthType string
+}
+
+// NewAuth creates a new auth filter specification to validate
+// authorization for requests. Current implementation uses Bearer
+// tokens to authorize requests, optionally check realms and
+// optionally check scopes.
+func NewAuth(o Options) filters.Spec {
+	cfg := oauth2.Config{
+		Endpoint: oauth2.Endpoint{
+			TokenURL: o.TokenURL,
+		},
+	}
+
+	return newSpec(typeForName(o.AuthType), cfg)
+}
+
+func typeForName(s string) roleCheckType {
+	switch s {
+	case AuthAllScopeName:
+		return checkAllScopes
+	case AuthAnyScopeName:
+		return checkAnyScopes
+	case AuthAnyKVName:
+		return checkAnyKV
+	case AuthAllKVName:
+		return checkAllKV
+	}
+	return checkUnknown
+}
+
+func (s *spec) Name() string {
+	switch s.typ {
+	case checkAnyScopes:
+		return AuthAnyScopeName
+	case checkAllScopes:
+		return AuthAllScopeName
+	case checkAnyKV:
+		return AuthAnyKVName
+	case checkAllKV:
+		return AuthAllKVName
+	}
+	return AuthUnknown
+}
+
+// CreateFilter creates an auth filter. All arguments have to be
+// strings. The first argument is the realm and the rest the given
+// scopes to check. How scopes are checked is based on the type. The
+// shown example will grant access only to tokens from `myrealm`, that
+// have scopes `read-x` and `write-y`:
+//
+//     s.CreateFilter("myrealm", "read-x", "write-y")
+//
+func (s *spec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+	if len(sargs) == 0 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+	if s.typ == checkUnknown {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	ac := &authClient{urlBase: s.cfg.Endpoint.TokenURL}
+
+	f := &filter{typ: s.typ, authClient: ac, kv: make(map[string]string)}
+	if len(sargs) > 0 {
+		switch f.typ {
+		case checkAnyKV:
+			fallthrough
+		case checkAllKV:
+			f.realm = sargs[0]
+			sargs = sargs[1:]
+			for i := 0; i+1 < len(sargs); i += 2 {
+				f.kv[sargs[i]] = sargs[i+1]
+			}
+			if len(sargs) == 0 || len(sargs)%2 != 0 {
+				return nil, filters.ErrInvalidFilterParameters
+			}
+		default:
+			f.realm, f.scopes = sargs[0], sargs[1:]
+		}
+
+	}
+
+	return f, nil
+
+}
+
+func (kv kv) String() string {
+	var res []string
+	for k, v := range kv {
+		res = append(res, k, v)
+	}
+	return strings.Join(res, ",")
+}
+
+// String prints nicely the filter configuration based on the
+// configuration and check used.
+func (f *filter) String() string {
+	switch f.typ {
+	case checkAnyScopes:
+		return fmt.Sprintf("%s(%s,%s)", AuthAnyScopeName, f.realm, strings.Join(f.scopes, ","))
+	case checkAllScopes:
+		return fmt.Sprintf("%s(%s,%s)", AuthAllScopeName, f.realm, strings.Join(f.scopes, ","))
+	case checkAnyKV:
+		return fmt.Sprintf("%s(%s,%s)", AuthAnyKVName, f.realm, f.kv)
+	case checkAllKV:
+		return fmt.Sprintf("%s(%s,%s)", AuthAllKVName, f.realm, f.kv)
+	}
+	return AuthUnknown
+}
+
+func (f *filter) validateRealm(h map[string]interface{}) bool {
```

Comment:
This will addressed as @aryszka suggested above, create 2 checks (new name style):

    oauthTokeninfoAnyScope("uid", "service-read")
    oauthTokeninfoRealmAnyScope("/employees", "uid", "service-read")